source/MaterialXContrib/MaterialXNode/MaterialXData.cpp

```diff
@@ -1,49 +1,102 @@
+{
+    // Create document
+    mx::DocumentPtr document = mx::createDocument();
+    if (!document)
+    {
+        throw mx::Exception("Failed to create a MaterialX document");
+    }
+
+    // Load libraries
+    static const mx::StringVec libraries = { "stdlib", "pbrlib", "bxdf", "stdlib/genglsl", "pbrlib/genglsl" };
+    MaterialXMaya::loadLibraries(libraries, librarySearchPath, document);
+
+    // Read document contents from disk
+    mx::XmlReadOptions readOptions;
+    readOptions.skipDuplicateElements = true;
+    mx::readFromXmlFile(document, materialXDocumentPath, mx::EMPTY_STRING, &readOptions);
+
+    return document;
+}
 }
 
-bool MaterialXData::setRenderableElement(const std::string& materialXDocument, const std::string& elementPath)
+MaterialXData::MaterialXData(   mx::DocumentPtr document,
+                                const std::string& elementPath, 
+                                const MaterialX::FileSearchPath& librarySearchPath )
+    : _document(document)
+    , _shaderGenerator(mx::GlslFragmentGenerator::create())
+    , _genContext(_shaderGenerator)
 {
-    clearXml();
-    _document = nullptr;
-    _element = nullptr;
+    if (!_document)
+    {
+        throw mx::Exception("No document specified");
+    }
+
+    std::vector<mx::TypedElementPtr> renderableElements;
+    mx::findRenderableElements(_document, renderableElements);
 
-    createDocument(materialXDocument);
-    if (_document)
+    // Nothing specified. Find the first renderable element and use that
+    if (elementPath.length() == 0)
     {
-        // Nothing specified. Find the first renderable element and use that
-        if (elementPath.length() == 0)
+        if (renderableElements.empty())
         {
-            std::vector<mx::TypedElementPtr> elements;
-            mx::findRenderableElements(_document, elements);
-            if (!elements.empty())
+            throw mx::Exception(
+                "No element path specified and no renderable elements in the document."
+            );
+        }
+
+        _element = renderableElements.front();
+    }
+    else
+    {
+        _element = _document->getDescendant(elementPath);
+
+        auto it = std::find_if(
+            renderableElements.begin(),
+            renderableElements.end(),
+            [this](mx::TypedElementPtr renderableElement) -> bool
             {
-                _element = elements[0];
-                return true;
+                return _element->getNamePath() == renderableElement->getNamePath();
             }
-        }
-        else
+        );
+
+        if (it == renderableElements.end())
         {
-            _element = _document->getDescendant(elementPath);
-            return isRenderable();
+            throw mx::Exception("The specified element is not renderable");
         }
     }
-    return false;
+
+    _genContext.registerSourceCodeSearchPath(librarySearchPath);
+
+    try
+    {
+        generateFragment();
```

Comment:
Calling shader generation from the constructor (and making it private), so that once a MaterialXData has been constructed without exceptions, it's supposed to store a valid fragment source.